src/extension.ts

```diff
@@ -18,24 +20,62 @@ export class RemixExtension {
     window.addEventListener('message', event => this.getMessage(event), false)
   }
 
+  /** Manage a message coming from the parent origin */
   private getMessage(event: MessageEvent) {
+    if (!event.source) return
+    if (!this.checkOrigin(event.origin)) return
     if (!event.data) return
-    if (typeof event.data !== 'string') return
     const msg = JSON.parse(event.data) as Message
     if (!msg) return
 
-    const { action, key, type, value } = msg
-    if (action === 'notification') {
-      if (this.notifications[key] && this.notifications[key][type]) {
-        this.notifications[key][type](value)
+    const { action, key, type, value, id, error } = msg
+    try {
+
+      if (action === 'request' && key === 'handshake') {

```

Comment:
an `handshake` requires at least 3 messages before the connection can be made no? seems only one is requested here

Reply:
Not sure to understand.
Here the handshake comes from the IDE on plugin activation. It enables the plugin to store the source window without using `parent.window`.
We can also add a handshake response from the plugin to the IDE if it solves a security leak, but I don't see which one right now.
Anyway, it would be maximum 2 messages no ?

Reply:
yes you're right, the third message would be final acknowledge from the parent to the plugin, but that's not mandatory.
Where is the handshake msg from plugin to parent in our case?

Reply:
There is none right now because I don't actually see what security leak it would solve...